Report errors using error_string in FollowJointTrajectory action result

miguelprada opened the pull request:
A couple of colleagues and me just spent a couple of hours troubleshooting a failing trajectory execution that was apparently caused by unsynchronized clocks causing trajectories being received by the controller with all points in the past.

It would have saved us a lot of time if some error message was displayed.

mathias-luedtke reviewed (dismissed):
Good catch!

bmagyar commented:
Should we try to capture these in the action interface too?
There's plenty of space left to extend the list error codes [here](http://docs.ros.org/lunar/api/control_msgs/html/action/FollowJointTrajectory.html).

mathias-luedtke commented:
> Should we try to capture these in the action interface too?

:+1: , but there is no need to add an error code. From the msg definition:
> INVALID_GOAL: The reason for the invalid goal (e.g., the requested
    trajectory is in the past).

~~`updateTrajectoryCommand` could be refactored to return the error code..~~
But we need to fill out the `error_string`. We could pass a string reference to `updateTrajectoryCommand`.

miguelprada commented:
Actually, the comment in the source code mislead me a little bit, but all points being in the past is not the only reason `initJointTrajectory` can return an empty trajectory. Therefore, the refactoring should affect this function as well.

~My suggestion would be to change the warning I just added to a more generic message about initializing the trajectory, and to open a new issue/PR to discuss/implement how the error reporting through the action interface could be improved.~

Changed my mind. It's not such a big undertaking. My only concern is how to handle the optional `InitJointTrajectoryOptions` argument to `initJointTrajectory`. Quick inspection suggests that it is only ever used (outside of the tests) by explicitly passing an options argument, but we never know whether removing the default value can cause regression in other packages. Otherwise, I could add an overload but it adds complexity without a clear need. What would you suggest?

miguelprada committed:
Return error string when failing to initialize trajectory from message

miguelprada commented:
I ended up adding an additional `std::string* error_string` argument to `updateTrajectoryCommand` and also as a field to `InitJointTrajectoryOptions`, so the signature of `initJointTrajectory` remains unchanged.

Let me know what you think.

mathias-luedtke commented:
> I ended up adding an additional std::string* error_string argument to updateTrajectoryCommand

That's what I proposed in a comment that I apparently forgot to actually send out :-/
In addition I wondered if [this warning](https://github.com/ros-controls/ros_controllers/blob/a9e650737d6d04d8a768c84a48763bdbc5e05800/joint_trajectory_controller/include/joint_trajectory_controller/init_joint_trajectory.h#L311) was not effective.

miguelprada commented:
Just fixed a [faulty line](https://travis-ci.org/ros-controls/ros_controllers/jobs/381558585#L647).

> In addition I wondered if this warning was not effective.

I was supporting my colleagues on the phone and I guess they were probably more focused on the machine running the application logic. You're right to point out that printing an additional message from the controller would not have helped much in this case, but sending the message back in the action result might improve the odds of future similar errors being caught faster.

mathias-luedtke commented on joint_trajectory_controller/include/joint_trajectory_controller/joint_trajectory_controller_impl.h:
```diff
@@ -555,7 +561,7 @@ updateTrajectoryCommand(const JointTrajectoryConstPtr& msg, RealtimeGoalHandlePt
     else
     {
-      // All trajectory points are in the past, nothing new to execute. Keep on executing current trajectory
+      ROS_WARN_NAMED(name_, "Failed to initialize trajectory from message, skipping.");
```
After actually reviewing `initJointTrajectory`, this warning seems to be superfluous.

miguelprada replied:
Agreed.

In retrospective, its clear that the issue was not really lacking an error message, but rather that attention was not being paid to the right terminal output. 😓

miguelprada committed:
Remove redundant warning messages

mathias-luedtke commented on joint_trajectory_controller/include/joint_trajectory_controller/joint_trajectory_controller_impl.h:
```diff
@@ -166,8 +166,13 @@ template <class SegmentImpl, class HardwareInterface>
   if (update_ok) {preemptActiveGoal();}
+  else
+  {
+    ROS_WARN_STREAM_NAMED(name_, "Failed to update trajectory: " << error_string);
```
redundant as well (?)

miguelprada replied:
Indeed.

miguelprada committed:
Report errors in updateTrajectoryCommand back though action result error_string

miguelprada commented:
I added changes to report back the potential errors in `updateTrajectoryCommand` using `error_string` as well.

Feel free to squash merge if you approve the changes.

Edit: apologies for all the amends and rebases. I’m working without a local build environment and its biting me hard...

mathias-luedtke commented:
> I’m working without a local build environment and its biting me hard...

https://github.com/ros-industrial/industrial_ci/pull/279 ;)

mathias-luedtke reviewed (approved):
I have resolved merge conflict and fixed a typo.

bmagyar commented on joint_trajectory_controller/include/joint_trajectory_controller/init_joint_trajectory.h:
```diff
@@ -308,10 +330,13 @@ Trajectory initJointTrajectory(const trajectory_msgs::JointTrajectory&       msg
+      error_string += "Last point is " + std::to_string(last_point_dur.toSec());
+      error_string += "s in the past.";
+      ROS_WARN_STREAM(error_string);
+      if (options.error_string)
```
Could we please use full brackets for these to match the format of the rest of the repo?

```
if (options.error_string)
{
  *options.error_string = error_string;
}
```

bmagyar replied:
Other than this it's all green lights from me too

mathias-luedtke committed:
implemented setErrorString

mathias-luedtke commented:
@bmagyar: Better?

bmagyar commented:
Thank you!!
